crates/bevy_ecs/src/system/commands/command_queue.rs

```diff
@@ -0,0 +1,167 @@
+use super::Command;
+use crate::world::World;
+
+/// # Safety
+///
+/// This function is only every called when the `command` bytes is the associated
+/// [`Commands`] `T` type. Also this only reads the data via `read_unaligned` so unaligned
+/// accesses are safe.
+unsafe fn invoke_command<T: Command>(command: *mut u8, world: &mut World) {
+    let command = if std::mem::size_of::<T>() > 0 {
+        command.cast::<T>().read_unaligned()
+    } else {
+        // NOTE: This is necessary because if the `CommandQueueInner` is only filled with 0-sized
+        // commands the `bytes` vec will never allocate. Then means that `bytes.as_ptr()` could be null
+        // and reading a null pointer is always UB.
+        // However according to https://doc.rust-lang.org/std/ptr/index.html
+        // "The canonical way to obtain a pointer that is valid for zero-sized accesses is NonNull::dangling"
+        // therefore the below code is safe to do.
+        let ptr = std::ptr::NonNull::<T>::dangling().as_ptr();
+        ptr.cast::<T>().read()
+    };
+    command.write(world);
+}
+
+struct CommandMeta {
+    offset: usize,
+    func: unsafe fn(value: *mut u8, world: &mut World),
+}
+
+#[derive(Default)]
+pub(crate) struct CommandQueueInner {
+    bytes: Vec<u8>,
+    metas: Vec<CommandMeta>,
+}
+
+// SAFE: All commands [`Command`] implement [`Send`]
+unsafe impl Send for CommandQueueInner {}
```

Comment:
```suggestion
unsafe impl Send for CommandQueueInner where Command: Send {}
```
Is this allowed?

Reply:
I'd be surprised

You could however make a compile_fail doctest where you male a command which isn't Send

Reply:
You could also use this to locally verify that all command are send: https://docs.rs/static_assertions/1.1.0/static_assertions/macro.assert_trait_sub_all.html
```rs
assert_trait_sub_all!(Command: Send, Sync);
```

Reply:
Seems that it is allowed: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=81217c580c0325fe7644bc8290ec7528

Reply:
Command is a trait: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=81217c580c0325fe7644bc8290ec7528

Reply:
My bad.

Reply:
I added a test that checks if command are `Send`
Basically:
```rust
fn assert_is_send(_: impl Send) {}
    fn a_command(command: impl Command) {
        assert_is_send(command);
    }
```